### PR TITLE
Byte size

### PR DESCRIPTION
**Pull request by lntel**
Please fill in this template.

- [✓] Use a meaningful title for the pull request. Include the name of the package modified.
- [✓] Test the change in your own code. (Compile and run.)
- [✓] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [✓] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [✓] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [✓] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [✓] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [✓] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [✓] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [✓] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [✓] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "dtslint/dt.json" }`, and no additional rules.
- [✓] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/byte-size/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// Type definitions for byte-size 8.1
+// Project: https://github.com/75lb/byte-size#readme
+// Definitions by: lntel <https://github.com/lntel>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export type byteSizeUnit = 'metric' | 'iec' | 'metric_octet' | 'iec_octet';
+
+export interface byteSizeOptions {
+    precision: number;
+    units: byteSizeUnit | string;
```
Comment (75lb) [inline types/byte-size/index.d.ts:10]: Remove the `| string` as it makes any string value valid here

File: types/byte-size/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    customUnits?: object;
+    toStringFn: () => void;
```
Comment (75lb) [inline types/byte-size/index.d.ts:12]: the `toStringFn` should return a `string`

**Review by 75lb — CHANGES_REQUESTED** [review]
Hi, thanks for the PR - first impressions attached..

**Review by 75lb — CHANGES_REQUESTED** [review]
I checked it some more and made more changes, this is what I ended up with.. does that look ok to you? 

```
export interface byteSizeOptions {
    precision?: number;
    units?: string;
    customUnits?: object;
    toStringFn?: () => string;
    locale?: string | string[];
}

export interface ByteSizeResult {
    value: string;
    unit: string;
    long: string;
    toString: () => string;
}

function byteSize(bytes: number, options?: byteSizeOptions): ByteSizeResult;    

namespace byteSize {
    export function defaultOptions (options: byteSizeOptions): void;
}

export default byteSize;
```

**Comment by sam-eah** [comment]
Hello, I may be wrong but shouldn't the decleration file look something like that ?
```ts
declare module "byte-size"{
    export interface byteSizeOptions {
        customUnits?: object;
        locale?: string | string[];
        precision?: number;
        toStringFn?: () => string;
        units?: string;
    }
    
    export interface ByteSizeResult {
        long: string;
        toString: () => string;
        unit: string;
        value: string;
    }
    
    function byteSize(bytes: number, options?: byteSizeOptions): ByteSizeResult;    
    
    namespace byteSize {
        export function defaultOptions (options: byteSizeOptions): void;
    }
    
    export default byteSize;
}
```

That's what I'm currently using for my project, it's working just fine!

**Comment by 75lb** [comment]
@sam-eah Yes, that is exactly how i'm using it locally too - with the types wrapped in `declare module "byte-size"`.. however, does it need to be written this way as a DT package? I'm not sure.. I checked other packages in the DT repo - some use `declare module` some don't..

**Comment by lntel** [comment]
Yeah I did start it like that but then read somewhere that with DT it's not necessary. I could be wrong though.

**Comment by 75lb** [comment]
> Yeah I did start it like that but then read somewhere that with DT it's not necessary. I could be wrong though.

Cool, once you've double-checked whether or not it's required could you also check the new types posted in the PR review? These types also include `byteSize.defaultOptions` and rename the result from `ByteSizeObject` to `ByteSizeResult`. 

Almost there, now..

**Comment by lntel** [comment]
> > Yeah I did start it like that but then read somewhere that with DT it's not necessary. I could be wrong though.
> 
> Cool, once you've double-checked whether or not it's required could you also check the new types posted in the PR review? These types also include `byteSize.defaultOptions` and rename the result from `ByteSizeObject` to `ByteSizeResult`.
> 
> Almost there, now..

So with DT it's not necessary to actually declare the module. I've made a couple changes too so give that a check.

File: types/byte-size/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+// Type definitions for byte-size 8.1
+// Project: https://github.com/75lb/byte-size#readme
+// Definitions by: lntel <https://github.com/lntel>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export type byteSizeUnit = 'metric' | 'iec' | 'metric_octet' | 'iec_octet';
```
Comment (75lb) [inline types/byte-size/index.d.ts:6]: I removed this line in my previous PR as the units is customisable..

File: types/byte-size/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+export interface byteSizeOptions {
+    precision?: number;
+    units?: byteSizeUnit;
```
Comment (75lb) [inline types/byte-size/index.d.ts:10]: i set this to `string` in my previous review.

File: types/byte-size/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    customUnits?: object;
+    toStringFn?: () => string;
+    locale?: string | string[];
+}
+
+export interface ByteSizeResult {
+    value: string;
+    unit: string;
+    long: string;
```
Comment (75lb) [inline types/byte-size/index.d.ts:19]: I added `toString` in my previous review..

File: types/byte-size/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+}
+
+declare function byteSize(bytes: number, options?: byteSizeOptions): ByteSizeResult;
+
+export function defaultOptions(options: byteSizeOptions): void;
```
Comment (75lb) [inline types/byte-size/index.d.ts:24]: I gave you the correct solution for this method in my previous review..

File: types/byte-size/byte-size-tests.ts [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+};
+
+byteSize(256, {
+    locale: 'en-gb',
+    precision: 1,
+    toStringFn,
+    units: 'iec_octet',
+});
+
+byteSize(12, {
```
Comment (75lb) [inline types/byte-size/byte-size-tests.ts:14]: The tests are inadequate and don't cover all methods, e.g. `byteSize.defaultOptions`..

**Review by 75lb — CHANGES_REQUESTED** [review]
I'm being as patient as possible but this PR has been broken from the start and remains broken, despite me giving you the complete solution in my previous review..

**Comment by 75lb** [comment]
The types file throws errors when I view it in VSCode.. 

Screenshot 2021-09-13 at 16 31 51

**Comment by lntel** [comment]
With the current solution you'd have to do the following 
image

And with the fix that you provided, there's an error as you cannot use namespace.

**Comment by 75lb** [comment]
I don't see any compile errors in VSCode or when I compile it on the command line.. 

Screenshot 2021-09-13 at 17 05 08

**Comment by 75lb** [comment]
> And with the fix that you provided, there's an error as you cannot use namespace.

I don't see an error.. the fix uses [Declaration Merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html).

**Comment by lntel** [comment]
But have you ran `npm test byte-size`? DT is failing it

**Comment by lntel** [comment]
image
This is my test with the code you wrote

**Comment by 75lb** [comment]
> This is my test with the code you wrote

So it needs to be within a `declare module` then?

```
declare module "byte-size"{
    export interface byteSizeOptions {
        customUnits?: object;
        locale?: string | string[];
        precision?: number;
        toStringFn?: () => string;
        units?: string;
    }
    
    export interface ByteSizeResult {
        long: string;
        toString: () => string;
        unit: string;
        value: string;
    }
    
    function byteSize(bytes: number, options?: byteSizeOptions): ByteSizeResult;    
    
    namespace byteSize {
        export function defaultOptions (options: byteSizeOptions): void;
    }
    
    export default byteSize;
}
```

**Comment by lntel** [comment]
Okay it was a strange fix, the module declaration didn't work so I had to find a workaround. Check it out now. My apologies.

File: types/byte-size/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,27 @@
+// Type definitions for byte-size 8.1
+// Project: https://github.com/75lb/byte-size#readme
+// Definitions by: lntel <https://github.com/lntel>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+export interface byteSizeOptions {
```
Comment (peterblazejewicz) [inline types/byte-size/index.d.ts:6]:
```suggestion
export interface ByteSizeOptions {
```
IMO better,

**Comment by peterblazejewicz** [comment]
This is OK and working, just this bit with non-standard naming convention, thx!

**Comment by 75lb** [comment]
Thanks.. i have finished my checks too - once the naming convention is fixed our work here is done..

**Review by 75lb — APPROVED** [review]

**Comment by 75lb** [comment]
thanks everyone 👍

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@lntel thx!

**Comment by lntel** [comment]
Ready to merge

**Pull request merged**